library/calendar.inc

```diff
@@ -15,30 +15,59 @@ function getIDfromUser($name)
     }
 }
 
-//  returns an array of facility id and names
-function getUserFacilities($uID)
+// Returns an array of the facility ids and names that the user is allowed to access.
+// Access might be for inventory purposes ($inventory=true) or calendar purposes.
+//
+function getUserFacilities($uID, $orderby = 'id', $inventory = false)
 {
-    if (!$GLOBALS['restrict_user_facility']) {
-        $rez = sqlStatement("
-		select id, name, color
-		from facility
-		where service_location != 0
-	");
+    $restrict = $inventory ? $GLOBALS['gbl_fac_warehouse_restrictions'] : $GLOBALS['restrict_user_facility'];
+    if ($restrict) {
+        // No entries in this table means the user is not restricted.
+        $countrow = sqlQuery(
+            "SELECT count(*) AS count FROM users_facility WHERE " .
+            "tablename = 'users' AND table_id = ?",
+            array($uID)
+        );
+    }
+    if (!$restrict || empty($countrow['count'])) {
+        $rez = sqlStatement(
+            "SELECT id, name, color FROM facility " .
+            "ORDER BY $orderby"
+        );
     } else {
-        $rez = sqlStatement("
-		  select uf.facility_id as id, f.name, f.color
-		  from users_facility uf
-		  left join facility f on (uf.facility_id = f.id)
-		  where uf.tablename='users' 
-		  and uf.table_id = ? 
-	    ", array($uID));
+        // This query gets facilities that the user is authorized to access.
+        $rez = sqlStatement(
+            "SELECT f.id, f.name, f.color " .
+            "FROM facility AS f " .
+            "JOIN users AS u ON u.id = ? " .
+            "WHERE f.id = u.facility_id OR f.id IN " .
+            "(SELECT DISTINCT uf.facility_id FROM users_facility AS uf WHERE uf.tablename = 'users' AND uf.table_id = u.id) " .
+            "ORDER BY f.$orderby",
```

Comment:
What are the possible values for $orderby?

Reply:
Searching for all calls to getUserFacilities() indicates it's always 'id'.

Reply:
to be safe here, would do:
```
"ORDER BY f." . escape_sql_column_name($orderby, 'facility'),
```

Reply:
Let's hold off on that. In principle $orderby might be any kind of ORDER BY clause and not necessarily a single column name. As things stand now it's fixed in code and not tainted by user input.